UnityProject/Assets/Scripts/Systems/Explosions/ExplosionComponent.cs

```diff
@@ -1,292 +1,57 @@
-			{
-				damagedObject.ApplyDamage(damage, AttackType.Bomb, DamageType.Burn);
-			}
-		}
-
-		private void DamageTiles(Vector3Int worldPosition, int damage)
-		{
-			var matrix = MatrixManager.AtPoint(worldPosition, true);
-			matrix.MetaTileMap.ApplyDamage(MatrixManager.WorldToLocalInt(worldPosition, matrix), damage, worldPosition, AttackType.Bomb);
-		}
-
-		/// <summary>
-		/// Plays explosion sound and shakes ground
-		/// </summary>
-		private void PlaySoundAndShake(Vector3Int explosionPosition, bool isEMP)
-		{
-			byte shakeIntensity = (byte)Mathf.Clamp(damage / 5, byte.MinValue, byte.MaxValue);
-			ExplosionUtils.PlaySoundAndShake(explosionPosition, shakeIntensity, (int)shakeDistance, isEMP);
-		}
-
-		private bool IsPastWall(Vector2Int pos, Vector2Int damageablePos, float distance)
-		{
-			return MatrixManager.RayCast((Vector2)pos, damageablePos - pos, distance, LayerTypeSelection.Walls , obstacleMask).ItHit == false;
-		}
-
-		private int CalculateDamage(Vector2Int damagePos, Vector2Int explosionPos)
-		{
-			float distance = Vector2Int.Distance(explosionPos, damagePos);
-			float effect = 1 - ((distance + distance) / ((radius + radius) + minDamage));
-			return (int)(damage * effect);
-		}
-
-		/// <summary>
-		/// calculates the distance from the the center using the looping x and y vars
-		/// returns a float between the limits
-		/// </summary>
-		private float DistanceFromCenter(Vector2Int pos, Vector2Int center, float lowLimit = 0.05f, float highLimit = 0.25f)
-		{
-			var dif = center - pos;
-
-			float percentage = (Mathf.Abs(dif.x) + Mathf.Abs(dif.y)) / (radius + radius);
-			float reversedPercentage = (1 - percentage) * 100;
-			float distance = ((reversedPercentage * (highLimit - lowLimit) / 100) + lowLimit);
-			return distance;
-		}
+		[TooltipAttribute("Explosion strength")]
+		public float strength = 50;
+
+		[Range(0, 150)]
+		[TooltipAttribute("Explosion radius. If it's equal to 0, it will be calculated using strength")]
+		public int radius = 0;
+
+		[Range(0, 255)]
+		[TooltipAttribute("Shaking strength. If it's equal to 0, it will be calculated using strength")]
+		public int shaking = 0;
+
+		[TooltipAttribute("Explosion type")]
+		public ExplosionType explosionType = ExplosionType.Regular;
+
+		public void SetExplosionData(float _strength, ExplosionType _explosionType = ExplosionType.Regular, int _radius = 0, int _shaking = 0)
+        {
+			if (0 > _shaking || _shaking > 255)
+            {
+				_shaking = 0;
+            }
+			if (0 > _radius || _radius > 150)
+            {
+				_radius = 0;
+            }
+
+			strength = _strength;
+			radius = _radius;
+			shaking = _shaking;
+			explosionType = _explosionType;
+        }
+
+		public void Explode()
+        {
+			Explosion.StartExplosion(GetComponentInChildren<Transform>().position.RoundToInt(), strength, nodeTypes[explosionType], radius, shaking);
+        }
+
+		public enum ExplosionType //add your explosion type here
```

Comment:
good candidate for https://refactoring.guru/es/replace-conditional-with-polymorphism

Reply:
cant open it. you dont like the dict?

Reply:
well, basically when you have an enum like this one it means somewhere in the code that uses the enum you will ask what type the explosion is and act accordingly. The refactoring techniques makes it so instead of that, you simply make subclasses for every time and tell the explosion to do its thing. Client code doesn't need to know anything about the implementation nor it needs to handle different types itself, it just tells the object to do the explosion however they do.

Reply:
so after talkin further we decided to leave it as it is, since there will be hardcode anyway so its possible to choose explosion type from editor. resolving it for now